frontend/test/metabase/lib/expressions/typeinferencer.unit.spec.js

```diff
@@ -0,0 +1,80 @@
+import { compile } from "metabase/lib/expressions/compile";
+import { infer } from "metabase/lib/expressions/typeinferencer";
+
+describe("metabase/lib/expressions/typeinferencer", () => {
+  function resolve(kind, name) {
+    return [kind, name];
+  }
+  function compileAs(source, startRule) {
+    let mbql = null;
+    try {
+      mbql = compile({ source, startRule, resolve });
+    } catch (e) {}
+    return mbql;
+  }
+
+  // workaround the limitation of the parsing expecting a strict top-level grammar rule
+  function tryCompile(source) {
+    let mbql = compileAs(source, "expression");
+    if (!mbql) {
+      mbql = compileAs(source, "boolean");
+    }
+    return mbql;
+  }
+
+  function type(expression) {
+    return infer(tryCompile(expression));
+  }
+
+  it("should infer the type of primitives", () => {
+    expect(type("0")).toEqual("number");
+    expect(type("1")).toEqual("number");
+    expect(type("3.14159")).toEqual("number");
+    expect(type('"Hola"')).toEqual("string");
+    expect(type("'Bonjour!'")).toEqual("string");
+  });
+
+  it("should infer the result of arithmetic operations", () => {
+    expect(type("[Price] + [Tax]")).toEqual("number");
+    expect(type("1.15 * [Total]")).toEqual("number");
+  });
+
+  it("should infer the result of comparisons", () => {
+    expect(type("[Discount] > 0")).toEqual("boolean");
+    expect(type("[Revenue] <= [Limit] * 2")).toEqual("boolean");
+    expect(type("1 != 2")).toEqual("boolean");
+  });
+
+  it("should infer the result of logical operations", () => {
+    expect(type("NOT [Deal]")).toEqual("boolean");
+    expect(type("[A] OR [B]")).toEqual("boolean");
+    expect(type("[X] AND [Y]")).toEqual("boolean");
```

Comment:
Just a note. We don't currently support boolean'ing like that, but I guess that's on your bucket list. Issue 12393

Reply:
Yup, this is right now just so we can have the type inference tested.